File: 2022/12/solution.ts
import * as fs from "fs";
import * as path from "path";
import MinHeap from "../../utils/MinHeap";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("").map((b) => b.charCodeAt(0)));

const directions = [
  [-1, 0],
  [1, 0],
  [0, -1],
  [0, 1],
];

const rMax = puzzleInput.length - 1;
const cMax = puzzleInput[0].length - 1;

let startPos: number[] = [];
let endPos: number[] = [];
let startFound = false;
let endFound = false;
for (let r = 0; !(startFound && endFound) && r <= rMax; r++) {
  for (let c = 0; !(startFound && endFound) && c <= cMax; c++) {
    if (puzzleInput[r][c] === "S".charCodeAt(0)) {
      startPos = [r, c];
      puzzleInput[r][c] = "a".charCodeAt(0);
      startFound = true;
    } else if (puzzleInput[r][c] === "E".charCodeAt(0)) {
      endPos = [r, c];
      puzzleInput[r][c] = "z".charCodeAt(0);
      endFound = true;
    }
  }
}

interface QueueEntry {
  steps: number;
  r: number;
  c: number;
}

function distanceToEnd(r: number, c: number): number {
  return Math.abs(r - endPos[0]) + Math.abs(c - endPos[1]);
}

function findMinSteps(startPositions: number[][]): number {
  const queue: [number, QueueEntry][] = startPositions.map((a) => [
    0,
    { steps: 0, r: a[0], c: a[1] },
  ]);
  const visited: Map<string, number> = new Map(
    startPositions.map((a) => [a.join(","), 0])
  );

  while (queue.length) {
    let [_, { steps, r, c }] = MinHeap.pop(queue) as [number, QueueEntry];
    if (r === endPos[0] && c === endPos[1]) {
      return steps;
    }
    steps++;
    const elev = puzzleInput[r][c];
    for (const [dr, dc] of directions) {
      const [r2, c2] = [r + dr, c + dc];
      if (r2 < 0 || c2 < 0 || r2 > rMax || c2 > cMax) continue;
      if (puzzleInput[r2][c2] - elev > 1) continue;
      if (visited.has(`${r2},${c2}`) && visited.get(`${r2},${c2}`)! <= steps)
        continue;
      visited.set(`${r2},${c2}`, steps);
      MinHeap.push(queue, [
        distanceToEnd(r2, c2) + steps,
        {
          steps,
          elev: puzzleInput[r2][c2],
          r: r2,
          c: c2,
        },
      ]);
    }
  }
  return -1;
}

function part1(): number {
  return findMinSteps([startPos]);
}

function part2(): number {
  const startPositions: number[][] = [];
  for (let r = 0; r <= rMax; r++) {
    for (let c = 0; c <= cMax; c++) {
      if (puzzleInput[r][c] === "a".charCodeAt(0)) {
        startPositions.push([r, c]);
      }
    }
  }
  return findMinSteps(startPositions);
}

console.log(part1());
console.log(part2());

File: 2022/12/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	heights, start, end := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(heights, start, end))
	fmt.Println(part2(heights, end))
}

type QueueEntry struct {
	steps int
	pos   [2]int
	ht    rune
}

var directions = [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

func parseInput(data []string) ([][]rune, [2]int, [2]int) {
	heights := make([][]rune, len(data))
	var start, end [2]int
	for i, row := range data {
		heightRow := make([]rune, len(row))
		for j, ch := range row {
			if ch == 'S' {
				start = [2]int{i, j}
				heightRow[j] = 0
			} else if ch == 'E' {
				end = [2]int{i, j}
				heightRow[j] = 'z' - 'a'
			} else {
				heightRow[j] = ch - 'a'
			}
		}
		heights[i] = heightRow
	}
	return heights, start, end
}

func part1(heights [][]rune, start, end [2]int) int {
	rMax := len(heights) - 1
	cMax := len(heights[0]) - 1

	queue := []QueueEntry{{0, start, 0}}
	visited := make(map[[2]int]struct{})
	visited[start] = struct{}{}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		if curr.pos == end {
			return curr.steps
		}
		for _, d := range directions {
			pos2 := [2]int{curr.pos[0] + d[0], curr.pos[1] + d[1]}
			if pos2[0] < 0 || pos2[1] < 0 || pos2[0] > rMax || pos2[1] > cMax {
				continue
			}
			ht := heights[pos2[0]][pos2[1]]
			if ht > (curr.ht + 1) {
				continue
			}
			if _, exists := visited[pos2]; exists {
				continue
			}
			visited[pos2] = struct{}{}
			queue = append(queue, QueueEntry{curr.steps + 1, pos2, ht})
		}
	}

	return math.MaxInt
}

func part2(heights [][]rune, end [2]int) int {
	var starts [][2]int
	for r, row := range heights {
		for c, ht := range row {
			if ht == 0 {
				starts = append(starts, [2]int{r, c})
			}
		}
	}

	minSteps := math.MaxInt
	for _, start := range starts {
		res := part1(heights, start, end)
		if res < minSteps {
			minSteps = res
		}
	}

	return minSteps
}
